File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    int n = 0;
    long long m = 0, r = 0, l = 0;
    vector<long long> tree;
    cin >> n >> m;
    for (int i = 0; i < n; i++)
    {
        int tmp;
        cin >> tmp;
        tree.push_back(tmp);
    }

    for (int i = 0; i < n; i++)
    {
        r += tree[i];
    }
    
    while (r >= l)
    {
        long long mid = 0;
        long long tmp = 0;
        mid = (r + l) / 2;
        
        for (int i = 0; i < n; i++)
        {
            if(mid < tree[i])
                tmp += (tree[i] - mid);
        }
        
        if(tmp == m)
        {
            r = mid;
            break;
        }

        if(tmp < m)
        {
            r = mid - 1;
        }
        else
        {
            l = mid + 1;
        }

        
    }
    
    cout << r;
    return 0;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
n , m = list(map(int,sys.stdin.readline().split())) #나무 개수 , 가져가는 나무 길이
tree= list(map(int,sys.stdin.readline().split()))
tree.sort()
h = 0

#높이 기준을 으로 나눈다
def bSherch(start,end):
    tmp = 0
    #start ,end 로 임시 높이를 구한다
    h = (start + end) // 2
    
    if start > end :   
        return h


    
    for i in range(n):
        if tree[i] > h:
            tmp += (tree[i] - h)
    if m == tmp:
        return h
    elif m > tmp:
        return bSherch(start,h - 1)
    elif m < tmp:
        return bSherch(h + 1,end)
    #임시 높이보다 높은 나무만 임시 높이로 빼고 총합을 구한다
    #높이랑 비교해서 리턴 or 재귀
    
print(bSherch(1,max(tree)))